### ISSUE TITLE
Data Driven Spiking Neural Network Optimization in Julia

### ISSUE THREAD

**Issue by russelljjarvis** [brainhackorg/global2021#98]
### Title

Data-Driven Spiking Neural Network Optimization in Julia

### Leaders

Russell Jarvis
@rjjarvis (mattermost)
@russelljjarvis (github)

### Collaborators

Seeking collaborators

### Brainhack Global 2021 Event

Brainhack Australasia

### What are you doing, for whom, and why?
- - The project is for everyone who wants to try to do data fitting of neuron models **without** Python+external simulation tools.
-- The project is for people who have been **frustrated** by writing support code in Python for externally supported simulator tools.
- - The project intends to consolidate neural ephys analysis tools and support that have existed in Julia for a while by applying tools by example.

### What makes your project special and exciting?
- -It has a basis in Julia-lang which might be novel and exciting for many people, yet, the scope of the project is meta-package. Meaning that many of the goals of the project are begginer friendly.
### How to get started?
Follow getting started here:
https://github.com/russelljjarvis/SpikeNetOpt.jl
### Link to project repository/sources

https://github.com/russelljjarvis/SpikeNetOpt.jl

### Goals for Brainhack Global

* Refactor optimizer design from bespoke specific examples to a general user interface. 
* Benchmarking of three different Julia Spiking Neural Network backend simulator speeds.
* eliminate global variables and untyped variables.

### Good first issues

1. issue one: make package installable via Project.toml, etc.

2. issue two: Refactor optimizer design from bespoke specific examples to a general user interface. 

3. make scatter plot animation of optimizer succeeding.

4. Use existing Python/BluePyOpt code to draw the GA evaluated error surface.

### Communication channels

https://mattermost.brainhack.org/brainhack/channels/spikenetoptjl

### Skills

- git push/pull
- git commit 
- git PR
- A desire to learn Julia (but with pre-existing programming knowledge)

### Onboarding documentation

I will create a contributing.md in the interim.

### What will participants learn?

* Julia package management of projects.
* Julia structs
* Julia multiple dispatch
* Julia typing

### Data to use

At the moment single-cell model fitting occurs, Allen Brain Observatory data, I have written a Python API included in the Julia code repository (python is called from Julia).  

https://observatory.brain-map.org/visualcoding/sdk/index

### Number of collaborators

4

### Credit to collaborators

I have started using the all contributors tool which makes all contributions to the repository visible (including raising issues via GitHub, or even ideas conveyed outside of GitHub). Also, I can write a reference letter for substantive contribution.

[![All Contributors](https://img.shields.io/badge/all_contributors-54-orange.svg?style=flat-square)](#contributors-)

### Image

Leave this text if you don't have an image yet.

### Type

coding_methods, documentation, method_development, visualization

### Development status

1_basic structure

### Topic

data_visualisation, neural_networks, reproducible_scientific_methods, single_neuron_models, other

### Tools

NWB

### Programming language

Julia

### Modalities

other

### Git skills

0_no_git_skills, 1_commit_push, 2_branches_PRs, 3_continuous_integration

### Things to do after the project is submitted.

- [ ] Add a comment saying: `Hi @Brainhack-Global/project-monitors: my project is ready!`
- [ ] Twitter-sized summary of your project pitch.

**Comment by Remi-Gau** [comment]
hey @russelljjarvis 

This is looking good already. :grin: 

Remember to ping us with a `@Brainhack-Global/project-monitors` when this is ready

**Comment by russelljjarvis** [comment]
Thanks, @Remi-Gau, not sure how much time is left before it is due. I thought I would clean up the base repository a bit, add a contributors file with standard FOSS contribution policies. 

Put files in predictable places for any newcomers.

**Comment by Remi-Gau** [comment]
You have quite a bit of time still, no worries! 
I don't think that Brainhack Australasia has any deadlines for project submission beyond "before the event starts": though for project review it will be better for us if we don't have to do it in the last 5 minutes before the event starts. :wink:

**Comment by russelljjarvis** [comment]
Wow thanks for the quick response Remi, Okay, I will try to leave a buffer of 48 hours before :).

**Comment by russelljjarvis** [comment]
Hi @Brainhack-Global/project-monitors: my project is ready!

**Comment by Remi-Gau** [comment]
hey @russelljjarvis 
I think that new comers will be happy to have a a bit more info on some of the first issues. Writing "good first issue" can be tricky though, but definitely adding a couple of extra links to external / additional doc or even some pointers to specific functions or line of code in the code base can help people get started faster.

**Comment by russelljjarvis** [comment]
@Remi-Gau For sure. My issues are just stubs, and my code doc currently lacks useful entry points. I will fix this soon :)

**Comment by russelljjarvis** [comment]
Hi @Remi-Gau, I have added a lot more detail and links to the Github issues today. I have also assigned issue labels, and added a contributing file.  

Should I move the more elaborate issues to this brainhacks template or will this bloat it out?

**Comment by Remi-Gau** [comment]
No this looks fine as it is on our side: will click the publish button.

I would suggest that any further changes you make should be done on the issues in your repo: as you say we can to keep things leaner in here. :wink:

**Comment by russelljjarvis** [comment]
Awesome. Thanks

**Comment by Remi-Gau** [comment]
If you mean the link to your project on the BHG2021 website, it is this one: https://brainhack.org/global2021/project/project_98/

Which is pretty much the content of your top comment.

**Comment by russelljjarvis** [comment]
Yes, everything makes sense after today. Thanks for your help. Let me know if you want help with other projects in this process for future brain hack events.
